rdr_service/dao/consent_dao.py

```diff
@@ -33,29 +38,81 @@ def _get_query_non_test_summaries(cls, session):
-        if end_date is None:
-            query = query.filter(
-                or_(
-                    ParticipantSummary.consentForStudyEnrollmentFirstYesAuthored >= start_date,
-                    ParticipantSummary.consentForCABoRAuthored >= start_date,
-                    ParticipantSummary.consentForElectronicHealthRecordsAuthored >= start_date,
-                    ParticipantSummary.consentForGenomicsRORAuthored >= start_date
+    def get_participants_with_unvalidated_files(cls, session) -> Collection[ParticipantSummary]:
+        """
+        Finds all participants that should have submitted a consent, but don't have the corresponding file validated
+        """
+
+        primary_consent = aliased(ConsentFile)
+        cabor_consent = aliased(ConsentFile)
+        ehr_consent = aliased(ConsentFile)
+        gror_consent = aliased(ConsentFile)
+        primary_update_consent = aliased(ConsentFile)
+        query = (
+            cls._get_query_non_test_summaries(session)
+            .outerjoin(
+                primary_consent, and_(
+                    primary_consent.participant_id == ParticipantSummary.participantId,
+                    primary_consent.type == ConsentType.PRIMARY
                 )
-            )
-        else:
-            query = query.filter(
+            ).outerjoin(
+                cabor_consent, and_(
+                    cabor_consent.participant_id == ParticipantSummary.participantId,
+                    cabor_consent.type == ConsentType.CABOR
+                )
+            ).outerjoin(
+                ehr_consent, and_(
+                    ehr_consent.participant_id == ParticipantSummary.participantId,
+                    ehr_consent.type == ConsentType.EHR
+                )
+            ).outerjoin(
+                gror_consent, and_(
+                    gror_consent.participant_id == ParticipantSummary.participantId,
+                    gror_consent.type == ConsentType.GROR
+                )
+            ).outerjoin(
+                primary_update_consent, and_(
+                    primary_update_consent.participant_id == ParticipantSummary.participantId,
+                    primary_update_consent.type == ConsentType.PRIMARY_UPDATE
+                )
+            ).outerjoin(
+                QuestionnaireResponse,
+                QuestionnaireResponse.participantId == ParticipantSummary.participantId
+            ).outerjoin(
+                QuestionnaireResponseAnswer,
+                QuestionnaireResponseAnswer.questionnaireResponseId == QuestionnaireResponse.questionnaireResponseId
+            ).outerjoin(
+                QuestionnaireQuestion,
+                QuestionnaireQuestion.questionnaireQuestionId == QuestionnaireResponseAnswer.questionId
+            ).outerjoin(
+                Code,
+                and_(
+                    Code.codeId == QuestionnaireQuestion.codeId,
+                    Code.value == PRIMARY_CONSENT_UPDATE_QUESTION_CODE
+                )
```

Comment:
Should these joins from `QuestionnaireResponse`, `QuestionnaireResponseAnswer`, `QuestionnaireQuestion`, and `Code` be inner joins instead of outer joins? The filter condition `Code.codeId.isnot(None)` requires that records exist through this join path.

Reply:
They'd need to be outer joins. The query should return any participant summaries that have submitted a consent but don't have the file for that consent validated yet. So a participant that hasn't responded to the PrimaryUpdate question but needs an EHR validated should be returned.

The check for whether a code was matched from this join is within an `or` clause, so the check for a code can fail but another condition could still cause the participant to be returned.

Reply:
I see, makes sense, thanks.